arrow-libs/resilience/arrow-resilience/src/commonMain/kotlin/arrow/resilience/ScheduleEither.kt

```diff
@@ -0,0 +1,30 @@
+import arrow.core.Either
+import arrow.core.raise.Raise
+import arrow.core.raise.either
+import arrow.resilience.Schedule
+import arrow.resilience.ScheduleStep
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.ensureActive
+import kotlin.time.Duration
+
+public suspend fun <Error, Result, Output> Schedule<Error, Output>.retry(
+  action: suspend Raise<Error>.() -> Result,
+): Either<Error, Result> {
+  var step: ScheduleStep<Error, Output> = step
+
+  while (true) {
+    currentCoroutineContext().ensureActive()
+    when (val result = either { action() }) {
```

Comment:
using `recover` here is probably nicer and won't result in unnecessary `Either` wrapping

Reply:
Another possibility is using fold, which gives a similar code pattern as matching on the result of the Raise.

Reply:
I'm not sure how to use recover here 🤔 The recover function would need to return `Result`.

either function is a shortcut for fold so it is possible to do something like this:
```kt
    fold(
      action,
      recover = {
        when (val decision = step(it)) {
          is Schedule.Decision.Continue -> {
            if (decision.delay != Duration.ZERO) delay(decision.delay)
            step = decision.step
          }

          is Schedule.Decision.Done -> return it.left()
        }
      },
      transform = { return it.right() },
    )
```
IMO it is less readable but anyway if we want to keep two signatures (retryRaise and retryEither) it will be simpler to do just:
```kt
public suspend inline fun <Error, Result, Output> Schedule<Error, Output>.retryRaise(
  @BuilderInference action: Raise<Error>.() -> Result,
): Either<Error, Result> = retryEither { either(action) }
```

Reply:
so at the end I used fold :)